pkgs/development/tools/parsing/antlr/4.nix

```diff
@@ -0,0 +1,157 @@
+        description = "Powerful parser generator";
+        longDescription = ''
+          ANTLR (ANother Tool for Language Recognition) is a powerful parser
+          generator for reading, processing, executing, or translating structured
+          text or binary files. It's widely used to build languages, tools, and
+          frameworks. From a grammar, ANTLR generates a parser that can build and
+          walk parse trees.
+        '';
+        homepage = "https://www.antlr.org/";
+        sourceProvenance = with sourceTypes; [ binaryBytecode ];
+        license = licenses.bsd3;
+        platforms = platforms.unix;
+      };
+    };
+
+    runtime = {
+      cpp = stdenv.mkDerivation {
+        pname = "antlr-runtime-cpp";
+        inherit version;
+        src = self.source;
+
+        outputs = [ "out" "dev" "doc" ];
+
+        nativeBuildInputs = [ cmake ninja pkg-config ];
+        buildInputs =
+          lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.CoreFoundation;
+
+        postUnpack = ''
+          export sourceRoot=$sourceRoot/runtime/Cpp
+        '';
+
+        meta = with lib; {
+          description = "C++ target for ANTLR 4";
+          homepage = "https://www.antlr.org/";
+          license = licenses.bsd3;
+          platforms = platforms.unix;
+        };
+      };
+    };
+  });
+
+in {
+  antlr4_11 = (
+    (mkAntlr {
+      version = "4.11.1";
+      sourceSha256 = "sha256-SUeDgfqLjYQorC8r/CKlwbYooTThMOILkizwQV8pocc=";
+      jarSha256 = "sha256-YpdeGStK8mIrcrXwExVT7jy86X923CpBYy3MVeJUc+E=";
+    }).extend (self: super: {
+      runtime.cpp = super.runtime.cpp.overrideAttrs {
+        cmakeFlags = [
+          # Generate CMake config files, which are not installed by default.
+          "-DANTLR4_INSTALL=ON"
+
+          # Disable tests, since they require downloading googletest, which is
+          # not available in a sandboxed build.
+          "-DANTLR_BUILD_CPP_TESTS=OFF"
+        ];
+      };
+    })
+  ).antlr;
+
+  antlr4_10 = (
+    (mkAntlr {
+      version = "4.10.1";
+      sourceSha256 = "sha256-Z1P81L0aPbimitzrHH/9rxsMCA6Qn3i42jFbUmVqu1E=";
+      jarSha256 = "sha256-QZSdQfINMdW4J3GHc13XVRCN9Ss422yGUQjTOCBA+Rg=";
+    }).extend (self: super: {
+      runtime.cpp = super.runtime.cpp.overrideAttrs (attrs: {
+        buildInputs = lib.optional stdenv.isLinux libuuid ++ attrs.buildInputs;
+        cmakeFlags = [
+          "-DANTLR4_INSTALL=ON"
+          "-DANTLR_BUILD_CPP_TESTS=OFF"
+        ];
+      });
+    })
+  ).antlr;
+
+  antlr4_9 = (
+    (mkAntlr {
+      version = "4.9.3";
+      sourceSha256 = "1af3cfqwk7lq1b5qsh1am0922fyhy7wmlpnrqdnvch3zzza9n1qm";
+      jarSha256 = "0dnz2x54kigc58bxnynjhmr5iq49f938vj6p50gdir1xdna41kdg";
+    }).extend (self: super: {
+      runtime.cpp = super.runtime.cpp.overrideAttrs (attrs: {
+        patchFlags = [ "-p3" ];
+        buildInputs = [ utf8cpp ]
+          ++ lib.optional stdenv.isLinux libuuid
+          ++ attrs.buildInputs;
+      });
+    })
```

Comment:
Instead of doing overlays we should add an extra input the to generic packages where we can append buildInputs.